packages/dd-trace/src/aiguard/sdk.js

```diff
@@ -0,0 +1,164 @@
+      }
+    }
+    return result
+  }
+
+  #isToolCall (message) {
+    return message.tool_calls || message.tool_call_id
+  }
+
+  #getToolName (message, history) {
+    // 1. assistant message with tool calls
+    if (message.tool_calls) {
+      const names = message.tool_calls.map((tool) => tool.function.name)
+      return names.length === 0 ? null : names.join(',')
+    }
+    // 2. assistant message with tool output (search the linked tool call in reverse order)
+    const id = message.tool_call_id
+    for (let i = history.length - 2; i >= 0; i--) {
+      const item = history[i]
+      if (item.tool_calls) {
+        for (const toolCall of item.tool_calls) {
+          if (toolCall.id === id) {
+            return toolCall.function.name
+          }
+        }
+      }
+    }
+    return null
+  }
+
+  evaluate (messages, opts) {
+    if (!this._initialized) {
+      return super.evaluate(messages, opts)
+    }
+    const { block = false } = opts ?? {}
+    return this._tracer.trace(AI_GUARD_RESOURCE, {}, async (span) => {
+      const last = messages[messages.length - 1]
+      const target = this.#isToolCall(last) ? 'tool' : 'prompt'
+      span.setTag(AI_GUARD_TARGET_TAG_KEY, target)
+      if (target === 'tool') {
+        const name = this.#getToolName(last, messages)
+        if (name) {
+          span.setTag(AI_GUARD_TOOL_NAME_TAG_KEY, name)
+        }
+      }
+      span.meta_struct = {
+        [AI_GUARD_META_STRUCT_KEY]: {
+          messages: this.#truncate(messages)
+        }
+      }
+      let response
+      try {
+        const payload = {
+          data: {
+            attributes: {
+              messages,
+              meta: this._meta,
+            }
+          }
+        }
+        response = await executeRequest(
+          payload,
+          { url: this._evaluateUrl, headers: this._headers, timeout: this._timeout })
+      } catch (e) {
+        throw new AIGuardClientError('Unexpected error calling AI Guard service', { cause: e })
+      }
+      if (response.status !== 200) {
+        throw new AIGuardClientError(
+          `AI Guard service call failed, status ${response.status}`,
+          { errors: response.body?.errors })
+      }
+      let action, reason, blockingEnabled
+      try {
+        const attr = response.body.data.attributes
+        if (!attr.action) {
+          throw new Error('Action missing from response')
+        }
+        action = attr.action
+        reason = attr.reason
+        blockingEnabled = attr.is_blocking_enabled ?? false
+      } catch (e) {
+        throw new AIGuardClientError(`AI Guard service returned unexpected response : ${response.body}`, { cause: e })
+      }
+      span.setTag(AI_GUARD_ACTION_TAG_KEY, action)
+      span.setTag(AI_GUARD_REASON_TAG_KEY, reason)
+      if (block && blockingEnabled && action !== ALLOW) {
+        span.setTag(AI_GUARD_BLOCKED_TAG_KEY, 'true')
+        throw new AIGuardAbortError(reason)
+      }
+      return { action, reason }
```

Comment:
(maybe as a follow-up) because exceptions/generating stack traces can be expensive, i wonder if it'd be nice to add some kind of soft-fail configuration (ie `aiguard.softFail`), such that if it's enabled this doesn't throw, but instead returns `block: true` as part of the result along with `{ action, reason }` and let users handle it that way.

i don't have a problem w the exceptions as-is, as i can see it being the users intention to handle errors from `evaluate` as hard exceptions, but still wanted to float the suggestion!

Reply:
The evaluate method receives an optional block parameter that just does that:

```ts
      evaluate(messages: Message[], opts?: { block?: boolean }): Promise<Evaluation>;
```

It's `false` by default and in this case it will not throw the exception but let the customer decided what to do with the evaluations.